Expose PluginViewFactory and meta-data flag for custom initialization

jonamireh opened the pull request:
## Summary
This PR aims to allow library consumers to inflate their own `HyperionPluginView` as well as customize when Hyperion's `ActivityLifecycleCallbacks` get registered (if at all). This allow for more customizability in launching the Hyperion menu; for example, it could now be embedded in a whole new Activity or a `BottomSheetFragment`, if you want.

## API Changes
A new `PluginViewFactory` is exposed from the `Hyperion` static class and is sourced from the `AppComponent`. It's implemented by the new `HyperionPluginView.Factory` class. The concrete implementation is largely copied from `InstallationLifecycleDelegate`. Since `HyperionPluginView` is `internal`, `PluginViewFactory` does not expose the concrete `HyperionPluginView` class as a return type.

`com.willowtreeapps.hyperion.core.enableOnStart` is a new `meta-data` flag that library consumers can add to their `AndroidManifest.xml` to prevent Hyperion from registering its `ActivityLifecycleCallbacks`. **To maintain backwards compatibility, this defaults to `true`.** Library consumers can use `enable` or `disable` from the `Hyperion` static class to register & unregister the same `ActivityLifecycleCallbacks`. It's mentioned in the documentation that calling these functions only affects _future_ Activities; the Hyperion menu cannot be opened for Activties that have already been launched. Calling `open` from the `Hyperion` static class will have no effect if the `ActivityLifecycleCallbacks` are not registered or were not executed for the foreground Activity.

### Documentation Updates
All of the relevant information for library consumers have been reflected in the README.

## Behavior Changes
`ApplicationInstaller#installIfNeeded` is now run exactly once when the ContentProvider binds the `Application` instance, as opposed to every `Activity#onCreate` of the app. This is internal, the change is transparent & backwards compatible

jonamireh committed:
Expose HyperionPluginView.Factory

Commit:
Add enableOnStart

handstandsam commented:
Thanks @jonamireh for making this PR.  This exposes the magic of Hyperion plugin discoverability without requiring activity view hierarchies to be modified at all.  Hopefully we can get this merged and published so that we can remove our own workarounds.

evant commented:
👍 I like the idea. However I'm seeing some odd behavior in the sample app that I'm worried are caused by these changes, mind taking a look? (tested on emulator running api 25)

https://user-images.githubusercontent.com/628930/162825879-0086050c-9877-420e-8aa7-287c5bf5f26d.mp4

Commit:
AppInfo metaData is nullable and HMC overrides onKeyListener

jonamireh commented:
I've addressed the bugs you've pointed out:
* `fragment_gesture.xml` has been wrapped in a `ScrollView`
  * Adjusted the padding as well
* `AppInfo#metaData` is actually nullable & needs to be checked on older API versions
  * This was an NPE before & caused Hyperion not to be enabled
* The implementation of `PluginViewFactory#create` no longer calls `HyperionMenuController#setPluginView` 
  * The `HyperionMenuController` configures an `OnKeyListener`, specifically for `KEYCODE_BACK`, to collapse the menu
  * This caused the `FactoryActivity` to never properly respond to back presses its `HyperionPluginView` from `PluginViewFactory` was swallowing all the relevant key events
  * A `package-private` `createInternal` allows binding to the `HyperionMenuController` for internal uses

Is it appropriate for the `PluginViewFactory#create` to not call `HyperionMenuController#setPluginView`? Do other plugins rely on this being called? I figure it only has to do with the open/close of the embedded menu but I wanted to be sure

Commit:
Code golf

Commit:
Update padding

evant commented:
Awesome changes! Though I think we've run into another issue. It doesn't seem like existing plugins play nicely with how embedding in a separate activity works for a couple of reasons:
1. Right now plugins are created when the activity is launched but destroyed when the activity is destroyed. That means you are unable to actually _use_ plugins that work outside of that activity. 
2. In addition some of the plugins assume you have a menu that opens and closes and listen to those callbacks, and behave badly without it.

The view inspecting plugins (Attributes Inspector & Measurement Inspector) are a good example of plugins that have problems. You can't actually use them on a target activity and they incorrectly enable themselves on the plugin view activity itself.

Will have to think through what our lifecycle should look like when embedding in a separate activity, open to suggestions here.

jonamireh commented:
I gave this a good amount of thought and I don't see an obvious way forward without ripping up the core internals of the project. Those plugins have an implicit dependency on being created & torn down with every Activity and this PR explicitly disables that functionality when using the PluginView as a standalone View. Effectively, you surrender the ability to have dynamic plugins based on the foreground Activity when you're in that mode

To make this work would probably mean reexamining what we really want to turn off with the `meta-data` field & maybe rework the foreground Activity tracking a little bit

Alternatively, we could accept this limitation, document it, & perhaps add an interface so dynamic plugins like this aren't even offered when creating the Plugin View as a standalone View

evant commented:
Yeah it would be helpful to know exactly what we want to get out of this, a couple of thoughts:

1. If we just wanted to support you customizing how the ui is presented but are fine with keeping the 'embedded in the activity approach' I think we could make that work without changing the api too much, may even be just allowing you to implement the menu interface yourself.
2. If it's super important it needs to be a separate activity, we could split out a simpler interface without view access that you could use. May even make sense to bundle an activity with the lib just for seeing those plugins that you could launch.

Commit:
Introduce isStandalone() to PluginModule

jonamireh commented:
> If it's super important it needs to be a separate activity, we could split out a simpler interface without view access that you could use. May even make sense to bundle an activity with the lib just for seeing those plugins that you could launch.

I made a small attempt at this by introducing a `boolean isStandalone()` method to `PluginModule`, which defaults to `true` to create an opt-out. I've gone ahead & taken a look at the first-party plugins that wouldn't be standalone & set their modules to return `false`

I had quite a bit of trouble understanding how the `PluginSource` is injected in the internal Dagger graph. The object is both set-able by the application but also is required for `AppComponent` dependencies, which seems at odds with each other. Instead, I added an internal `PluginFilter` that's consumed by the `HyperionPluginView` to filter out embed-only plugins when called from the public factory.

handstandsam commented:
Thanks @jonamireh for making these updates.  Hopefully we can get this in.

evant commented on hyperion-plugin/src/main/java/com/willowtreeapps/hyperion/plugin/v1/PluginModule.java:
```diff
@@ -48,6 +48,10 @@ protected void onDestroy() {
 
     }
 
+    public boolean isStandalone() {
```
Can you add some javadoc on what this method does?

evant commented on hyperion-sample/src/main/java/com/willowtreeapps/hyperion/sample/StandaloneActivity.java:
```diff
@@ -12,7 +12,7 @@
 
 @HyperionIgnore
-public class FactoryActivity extends AppCompatActivity {
+public class StandaloneActivity extends AppCompatActivity {
```
What do you think about moving this into hyperion?

Commit:
Add javadoc, move StandaloneActivity

evant commented:
Published a SNAPSHOT with these changes. https://oss.sonatype.org/content/repositories/snapshots/com/willowtreeapps/hyperion/hyperion-core/0.9.35-SNAPSHOT/

lmk if this solves your use-case and we can get out a proper release

handstandsam commented:
Thanks @evant!  Hopefully @jonamireh can test soon. Thanks for the reviews and for publishing!
